### ISSUE TITLE
Ability to track non-tracing scripts

### ISSUE THREAD

**Issue by SeungheonOh** [KtorZ/aiken-analytics#1]
From my very brief go-over of the code, it looks like it's only looking for some of the framework-specific trace messages on traced scripts. For more accurate metric, we can add some kind of small markers onto each script all scripts. Something as small as `\ctx -> (\_ -> (actual script)) "Plutarch"` will be enough to track both traced and un traced scripts. Cost of one extra lambda abstraction will be near negligible. The actual "tag" can be chosen more carefully so that it can be as small and with minimal overlap. 

At least for frameworks that implement something similar, we can grab more accurate metric.

**Comment by KtorZ** [comment]
Note that we have such marker for both Aiken and Plu-ts (since recently). 

- Aiken uses  `delay [ (error) (force (error)) ]` 
- Plu-ts uses: `(case (constr 0 ) <actual contract> (con integer 42))` (still have to integrate that one)

For other framework, that's indeed a best effort based on framework-specific trace messages and structure; through the frontier between Plutarch and Plutus-Tx (especially with Plutonomy) is quite blurry. So, more than happy to see more framework supporting a compiler-specific marker.

**Comment by SeungheonOh** [comment]
Ah neat. I'll try to get some clear marker in on the next release. This would've been nicer if we made some unified format for the markers but I guess this works too

**Comment by KtorZ** [comment]
A CIP is still an option. I talked to the Plutus Core team a while back about adding a 1 byte marker as part of the header for UPLC program  reserved to compilers (e.g. instead of the 3 digit Plutus version that has arguably any benefit 😬 ...); but the idea was rejected.

**Comment by KtorZ** [comment]
Also note that with the way Michele did his on Plu-ts, we can now simply all follow the same approach and reserve specific numbers for each compiler. (42 being now, Plu-ts).

**Comment by SeungheonOh** [comment]
```hs
(\qsrxlF0Y7 xF -> 10) 1 5
CPU: 112100
MEM: 800
SIZE: 13
 
(Case ((Constr 0 ()) ((\xF -> 10)) (1))) 5
CPU: 96100
MEM: 700
SIZE: 15
```
Ah neat, sop is slightly more cheaper. I thought wrapping in lambda would be cheaper. I feel like this is negligible enough to be not bother to do a dedicated CIP for. Michele's way looks pretty neat. We can just follow it I think.


edit: if anyone want 42, you can still have it : D
```
(Case ((Constr 0 ()) ((\xF -> 10)) (addInteger 4 2))) 5
CPU: 96100
MEM: 700
SIZE: 19
```

**Comment by KtorZ** [comment]
I guess for what it's worth, I just have us all commit to a bytestring identifier 😅 ...! 

```
case ((constr 0 ()) (...) (con bytestring "Aiken")

case ((constr 0 ()) (...) (con bytestring "Plutarch")

case ((constr 0 ()) (...) (con bytestring "Plu-ts")
```

etc..

**Comment by KtorZ** [comment]
cc @christianschmitz @nielstron @nau @michele-nuzzi

**Comment by nielstron** [comment]
is this optimized away by uplc level optimizers (i.e the one in Aiken)? should maybe add a case there to preserve the string or replace it with "... <optimizer>"

**Comment by KtorZ** [comment]
Good point @nielstron, it probably would, unless we make optimizers aware of it. Which should be pretty straightforward as this should be the first node we ever encounter in the program. Although there's also the case of parameterized validators, for which we could, by convention, always apply _below_ the "compiler header". This would mean that for analytics, we wouldn't even need to deserialise the entire program but merely the first node.

**Comment by nau** [comment]
Why not simply `[(lam compiler validator) (con string "Scalus")]`?

That would work for Plutus V1/V2, not just only V3. It's extra 3 nodes, 300 memory, 48000 cpu. Same as constr/case but more versatile.

**Comment by KtorZ** [comment]
I am happy with both @nau. I don't see reasons to compile to V1/V2 now that V3 is available, so I am not to bothered with the marker to be a V3-exclusive features (if anything, it might encourage everyone to upgrade :)). But, the lambda also works well, and since it's a one time cost; the benefits of a more "optimized" call here are really negligible.

**Comment by nau** [comment]
I'll add this to Scalus then

**Comment by michele-nuzzi** [comment]
>  I just have us all commit to a bytestring identifier 

@KtorZ if we are going to use it as standard I will comply, is not a big deal.

however I would still prefer to use constant integers due to how bad bytestring encoding in comparison.

But I repeat, we are talking about details

**Comment by michele-nuzzi** [comment]
> Why not simply `[(lam compiler validator) (con string "Scalus")]`?

One thing I don't like about this is that is not semantically null.

If in an hypothetical future we allow "inline uplc" (just like solidity has inline assembly) (that could turn useful for shared libs), people might want to use variables outside the scope assuming some additional parameter application offchain.

is not safe, and in theory only expert would use it, but theory is not reality.

**Comment by nau** [comment]
> however I would still prefer to use constant integers due to how bad bytestring encoding in comparison.

Why not use the `string` datatype? It's literally intended for strings :)
Saving few bytes doesn't makes much difference, does it?

> > Why not simply `[(lam compiler validator) (con string "Scalus")]`?
> 
> One thing I don't like about this is that is not semantically null.
> If in an hypothetical future we allow "inline uplc" (just like solidity has inline assembly) (that could turn useful for shared libs), people might want to use variables outside the scope assuming some additional parameter application offchain.
> 
> is not safe, and in theory only expert would use it, but theory is not reality.

OK

**Comment by michele-nuzzi** [comment]
Yes it is not a big difference

however these markers (and UPLC in general) are not meant to be human-friendly, main purpose is to detect them offchain, necessarily using tools.

so I don't see why use those extra bytes to make us humans happy

**Comment by christianschmitz** [comment]
@SeungheonOh Is the use-case for this important?

**Comment by SeungheonOh** [comment]
It's less about use case and more about allowing more accurate data collection. Which answer will vary depending on who you ask. I think having accurate measure of onchain framework usage is helpful

**Comment by nielstron** [comment]
@michele-nuzzi int *is* extra bytes, because int is at least 4 bytes. choosing bytestring "P" for Plu-Ts would be smaller

@nau (byte)strings are just pretty integers 

I recommend to toss a coin

**Comment by michele-nuzzi** [comment]
A constant integer under 7 bits is 18 bits in flat

4 const node
6 for the type
8 bit zigzag encoded integers

If I remember correctly

A single letter byte is between 18 and 24 bits due to padding

Also a single letter leaves space for ambiguity "P" is "Plutarch" or "Plu-ts" or "Plutus TX"?

**Comment by SeungheonOh** [comment]
We can just assign one character arbitrary. a for plutus, b for Aiken, ...

With enough coordination this could work. 

I don't think the size difference is any significant at all though . So as long as each framework picks different "stuff" it should do it

**Comment by michele-nuzzi** [comment]
As @nielstron stated

> (byte)strings are just pretty integers

**Comment by christianschmitz** [comment]
Although most proposals above indeed have a small impact on script size/budget, I think it would be much better to put this effort into a datum standard for reference scripts.

We can assume that most serious smart contracts use reference scripts. The datums of those UTxOs can be used to provide much richer information: which compiler/version was used, commit hash of original code, github URL (or ipfs etc.)

Such a standard could even give an additional incentive to projects to open-source their smart contract code.

Then explorers and other tools can automatically verify that the on-chain code indeed corresponds to the original code, source-mapping can be automatically applied, validators created with compiler versions which contain known bugs/exploits can be flagged, etc. etc.

**Comment by christianschmitz** [comment]
I would also like to point out that things that seem "negligible" and "not significant" now, could become significant in the future.

We all want Cardano mainnet usage to go up, which means it will become congested and fees will rise. We can do a back-of-the-envelope cost-benefit analysis by assuming a future ADA marketcap of 100 BUSD, fully supported by fees and no further growth perspective.

With an ROI of 5% that equates to roughly 5 BUSD fees per year income. Let's further assume 50% of this is generated by executing scripts and that the annual execution budget stays the same (mem: 50_000_000 . 365 . 24 . 60 . 3 = 78.84e12, cpu: 40_000_000_000 . 365 . 24 . 60 . 3 = 6.3072e16). Let's further assume that mem and cpu each account for half the fee, we end up with about 1.5e-5 USD/mem and 2e-8 USD/cpu

Assuming 3 UPLC nodes are added, so mem +300 and cpu +48000: 0.005 USD extra cost per validation. Assuming on average two validations per smart contract transaction, assuming 25% of the future transactions involve smart contracts and 10 TPS: 0.005 USD/validation . 2 validations . 25% . 10 TPS . 3600 . 24 . 365 = **788400 USD/year**.

So we'd collectively be spending almost 1 MUSD on 3 additional UPLC nodes, for something that will only have a small benefit.

These are of course very rough estimates, and this calculation might even contain a mistake. But we should probably refrain from using the words "negligible" and "not significant" when talking about UPLC optimization.

**Comment by SeungheonOh** [comment]
I agree this tagging isn't most elegant nor efficient. But other option would be CIP as @KtorZ said. Honestly, I saw more involvement and support in this thread than I expected so perhaps CIP wouldn't be too difficult assuming similar support we saw here . 

If we take that route. I think best way is to have non-executing data in encoded UPLC--CBOR encoded Data, bytestring, integer, I don't know. Basically same as that three node trick but this won't actually get evaluated. That way we track all scripts instead of just reference scripts. I disagree the statement that all serious smart contracts are reference script.

**Comment by christianschmitz** [comment]
A non-executing extra byte in the UPLC-CBOR would be much cheaper (about 30x), but would still incur a global cost of several 10 kUSD/year due to how ref-script size impacts the tx-fee.

**Comment by KtorZ** [comment]
>  @christianschmitz: but would still incur a global cost of several 10 kUSD/year due to how ref-script size impacts the tx-fee.

Not sure where you're pulling the 10kUSD/year cost (in particular, I think you are 2 orders of magnitude off 😅) 

The marker above would cost about 100 lovelace (`3*15` of ref script, `800*0.0577` of mem, `112100*7.21e-5` of cpu) per transaction/execution; with a rate of about 150k transactions per month on average; that's about 180 Ada a year 🤔 

> @christianschmitz: I think it would be much better to put this effort into a datum standard for reference scripts.

That moves the problem to dApp developers, which has already proven to not work for enforcing behaviors. First because people need to be aware of the standard, and second, because most will still refuse to clutter their smart contract flow with an extra piece of information. 

---- 

I also don't want to over do it guys. My main goal with this project is to provide somewhat reliable analytics about the on-chain traffic. So, all I am really after is for a way to classify smart contract activity. Which means that **we do not need to reach consensus on the approach**. If @christianschmitz you prefer encouraging people to use a Datum and makes this the de-facto method for tracking Helios contract, so be it, I'll take it! So long as we have one method for each, I am happy.

**Comment by christianschmitz** [comment]
@KtorZ at current prices and at current load yes. But usage and fees must increase dramatically for the network to be sustainable. 150 k (smart contract) transactions per month is very low. I was assuming closer to 200 k per day

**Comment by KtorZ** [comment]
I like your bullish mindset @christianschmitz 😅

**Comment by nau** [comment]
I've made some calculations:
<img width="827" alt="image" src="https://github.com/user-attachments/assets/bd52fbd7-f760-4a17-861b-bcad5af13195">

and got these number:

    extraNodes: 3, fee cpu: 3.4608000000000003, fee memory: 8.655, total fee: 12
    numTxPerYear: 315360000, total fees: 3784 Ada

So if even ALL theoretically possible transactions (315 millions!) would use extra 3 nodes we'd get 3700 extra ada in fees. Realisticly, that will be at least 50 times less. So, it's ok.

Please note, that current node memory usage of 100 is insane and I will initiate a protocol parameter change to lower it at least to 50. And `Var` memory usage should be 1, ideally.

Also, expect cpu/memory prices to become lower as well.

That said, I'm totally fine with extra 3 nodes.

**Comment by KtorZ** [comment]
You had to bring Scala into this @nau, didn't you 😅! Yet, yes; I don't think the cost is that big of a problem. So please, everyone, let me know which method your prefer for tracking contracts. So far, we have:

| Framework | Marker |
| --- | --- | 
| `Aiken` | `delay [ (error) (force (error)) ]` |
| `Plu-ts` | `(case (constr 0) validator (con integer 42))` |
| `Plutarch` | `[(lam compiler validator) (con string "Plutarch")]` |
| `Scalus` | `[(lam compiler validator) (con string "Scalus")]` |

... ?

**Comment by christianschmitz** [comment]
@nau Absolute fees (in USD equivalent) should be way higher though, a little more than Ethereum currently. A successful network is an expensive network.

@KtorZ Helios will opt for the ref script datum. I'll think of a good standard so it can scale to other frameworks later on. Some scripts approach the limit for size/ex-budget, so every byte/UPLC-node really does count.

The ref script datum approach will take time to implement, but because Helios users tend to use the builtin ref script registry mechanism, it's easy to make it the default setting if the validator repo contains an open source license and is publicly hosted.

For non-ref Helios scripts: you can assume that a transaction after a certain date, lacking a marker, is one of the remaining compilers. If I'm the only that remains a stickler about this, then that should be only Helios.

**Comment by SeungheonOh** [comment]
> A non-executing extra byte in the UPLC-CBOR would be much cheaper (about 30x), but would still incur a global cost of several 10 kUSD/year due to how ref-script size impacts the tx-fee.

oof, forgot they changed the fee mechanize to take ref script size into consideration : /

**Comment by MicroProofs** [comment]
> You had to bring Scala into this @nau, didn't you 😅! Yet, yes; I don't think the cost is that big of a problem. So please, everyone, let me know which method your prefer for tracking contracts. So far, we have:
> 
> Framework	Marker
> `Aiken`	`[ (error) (force (error)) ]`
> `Plu-ts`	`(case (constr 0) validator (con integer 42))`
> `Plutarch`	`[(lam compiler validator) (con string "Plutarch")]`
> `Scalus`	`[(lam compiler validator) (con string "Scalus")]`
> ... ?

It's really actually a marker without the delay. The idea is the [error (can be any chosen marker)] is really small. And it doesn't get optimized out.

**Comment by MicroProofs** [comment]
The other benefit is that has a negligible size impact. 2 bytes vs .5 bytes and it has one node impact on the mem and cpu cost. That's why I think the [error (special marker)] is the best for script markers considering it has the smallest impact.

Aiken [ (error) (force (error)) ]
Plu-ts (case (constr 0) validator (con integer 42))
Plutarch [error (lam x error)]
Scalus [error (builtin addInt)]

**Comment by MicroProofs** [comment]
> A non-executing extra byte in the UPLC-CBOR would be much cheaper (about 30x), but would still incur a global cost of several 10 kUSD/year due to how ref-script size impacts the tx-fee.

This is true, so It's definitely a consideration.

**Comment by michele-nuzzi** [comment]
> I think it would be much better to put this effort into a datum standard for reference scripts.

@christianschmitz I use the same utxo to deploy the ref script AND have a constant datum, usually working as config, to reduce the number of ref scripts in a transaction (reducing the length of the ref script list for on-chain execution)

not sure a datum standard is the best, but if we are going for this ref script approach then metadata would be more appropriate.

Even though I'm with @KtorZ here since this approach is not enforceable.

**Comment by SeungheonOh** [comment]
> It's really actually a marker without the delay. The idea is the [error (can be any chosen marker)] is really small. And it doesn't get optimized out.

Application from and to error is automatically error. This is one of the easiest optimization strategy. We optimize this out for Plutarch. It can be disabled for certain cases, I suppose, but I don't think stand alone uplc optimizers would leave this alone.

This does sound like most efficient marker since it won't incur fixed cost increase on script, and we can ask optimizers to leave first error application statement of this case alone so that it won't get optimized away

**Comment by nau** [comment]
Gentlement, 

What we're trying to do here is to add "dead code" markers. So, unless we make special effort, all these markers can/should be eliminated by a sufficiently effective optimizer.
That said, I suggest us to expect these markers to be added only after all of the optimizations.

@MicroProofs, could you help me understand where the validator here? 
> Aiken [ (error) (force (error)) ]

If it's a marker then it should be put somewhere somehow, and it's more nodes then, right?

Also, could someone please clarify what do you mean by non-executing byte in CBOR encoded UPLC?
The node expects a byte array in the validator CBOR and I don't know how to add extra anything there.

> You had to bring Scala into this @nau, didn't you 😅!

@KtorZ, always! 🤣

**Comment by SeungheonOh** [comment]
Non-executing bytes would require CIP to add non-removable arbitrary bytes at the front of the script. Similar to UPLC version number they embed on front.

Error tags could be added after optimization step, by replacing first error term in traversal of each node.

I see another problem though, this would be not be able to mark scripts without erroring term. It's less of a problem since most(basically any scripts that isn't always validate script) but that might change since there's CIP-??? That proposes to add return value to validation.

**Comment by KtorZ** [comment]
> @nau: could you help me understand where the validator here? 

Every Aiken-produced UPLC program ends with an if-then-else expression, where the condition is the (fully saturated) validator, the then simply yields unit, and the else errors. But instead of simply erroring; we error out with this special marker since it is semantically equivalent. 

That's also why it doesn't get optimized in our case; it's out of the optimization pass since we only optimized the validator, but not the outer wrapper.

**Comment by nau** [comment]
How about this. 
We make a CIP to change the meaning of the 3rd, `patch`, number of Plutus version to be used by compilers. 

- This should be backwards/forwards compatible
- Won't impact script size
- Won't impact execution and fees.

Win-win.

**Comment by nau** [comment]
It's gonna be quite a large diff in the `plutus` repo though, at least on a first glance...

**Comment by nau** [comment]
I have an idea. 

Plutus V3 requires successful scripts to return `(con unit ())` according to [CIP-117](https://cips.cardano.org/cip/CIP-0117).
This means we could return a non-unit constant as an error that identifies the compiler!

So, instead of `(force (if validator (delay (con unit)) (delay error)))` we could generate
`(if validator (con unit) (con 1/2/"Plutarch"/42/etc))`.

This removes 4 nodes (16 bits) and removes `delay` and `force` evaluation, adds 18 bits for numbers between -63..+63.

This way we'll get the marker, lower execution costs, and basically the same script size.
It's even better than the "version" version!

Or even better. We could claim a range of number, say 10-19, that would indicate a compiler and its version. Scalus 0.x is 10, Scalus 1.0 is 11 etc.

This doesn't handle 'if-less' scripts that only rely on `builtin` failures, but those scripts are usualy small so we can ignore them.

What do you think?
@KtorZ @christianschmitz @nielstron @SeungheonOh @michele-nuzzi @MicroProofs

**Comment by SeungheonOh** [comment]
Main problem I see is not all script will have this "non-unit returns" as you mentioned. Some builtins are effectful and scripts may rely on those builtins to fail scripts instead of having explicit non-unit returns. For example, to check if Data is Integer, they would just call the built in `asInt` which will not have return, but will fail if given value is not a integer datum. 

Being said, I haven't checked how often "non-unit return" is used to fail scripts. But I don't recall using it often even for larger script. I'm not sure how it is for other projects

**Comment by nau** [comment]
No, no, no, we don't check the actual result of the evaluatation :)

Similar to Aiken approach I suggest just to search for the first `if _ then () else (con 1/2/3)` and if it's there – we identified the compiler.
All we need to do is compilers to generate this instead of `if _ then () else (error)` for the validator wrapper when possible.

**Comment by gufmar** [comment]
Some months later I try to understand how you ended up on this issue. 

Is it worth looking into dbsync.script.bytes field 
decode it 
and look for some of the discussed patterns/markers?

if you started placing such markers (?), it would be useful to have one TX id per language and compiler, I can look into for practical patterns...
